clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -3454,9 +3454,11 @@ void Sema::DeclareGlobalAllocationFunction(DeclarationName Name,
       if (Func->getNumParams() == Params.size()) {
         llvm::SmallVector<QualType, 3> FuncParams;
         for (auto *P : Func->parameters())
-          FuncParams.push_back(
-              Context.getCanonicalType(P->getType().getUnqualifiedType()));
-        if (llvm::ArrayRef(FuncParams) == Params) {
+          FuncParams.push_back(P->getType().getUnqualifiedType());
+        if (std::equal(FuncParams.begin(), FuncParams.end(), Params.begin(),
+                       Params.end(), [&](QualType LT, QualType RT) {
+                         return Context.hasSameType(LT, RT);
```

Comment:
Actually, you can just drop the temporary vector, and just compare with hasSameUnqualifiedType. That's more formally correct, would allow to use this function with top level qualifiers.